lib/connection.cpp

```diff
@@ -1771,3 +1890,195 @@ QVector<Connection::SupportedRoomVersion> Connection::availableRoomVersions() co
+        currentQueryKeysJob = nullptr;
+    }
+    auto queryKeysJob = q->callApi<QueryKeysJob>(users);
+    currentQueryKeysJob = queryKeysJob;
+    connect(queryKeysJob, &BaseJob::success, q, [this, queryKeysJob](){
+        currentQueryKeysJob = nullptr;
+        const auto data = queryKeysJob->deviceKeys();
+        for(const auto &[user, keys] : asKeyValueRange(data)) {
+            deviceKeys[user].clear();
+            for(const auto &device : keys) {
+                if(device.userId != user) {
+                    qCWarning(E2EE) << "mxId mismatch during device key verification:" << device.userId << user;
+                    continue;
+                }
+                if(!device.algorithms.contains("m.olm.v1.curve25519-aes-sha2") || !device.algorithms.contains("m.megolm.v1.aes-sha2")) {
+                    qCWarning(E2EE) << "Unsupported encryption algorithms found" << device.algorithms;
+                    continue;
+                }
+                if(!verifyIdentitySignature(device, device.deviceId, device.userId)) {
+                    qCWarning(E2EE) << "Failed to verify devicekeys signature. Skipping this device";
+                    continue;
+                }
+                deviceKeys[user][device.deviceId] = device;
+            }
+            outdatedUsers -= user;
+        }
+        saveDevicesList();
+    });
+}
+
+void Connection::encryptionUpdate(Room *room)
+{
+    for(const auto &user : room->users()) {
+        if(!d->trackedUsers.contains(user->id())) {
+            d->trackedUsers += user->id();
+            d->outdatedUsers += user->id();
+            d->encryptionUpdateRequired = true;
+        }
+    }
+}
+
+void Connection::Private::saveDevicesList()
+{
+    if (!cacheState)
+        return;
+
+    QElapsedTimer et;
+    et.start();
+
+    QFile outFile { q->e2eeDataDir() % "/deviceslist.json" };
+    if (!outFile.open(QFile::WriteOnly)) {
+        qCWarning(E2EE) << "Error opening" << outFile.fileName() << ":"
+                        << outFile.errorString();
+        qCWarning(E2EE) << "Caching the rooms state disabled";
+        cacheState = false;
+        return;
+    }
+
+    QJsonObject rootObj {
+        { QStringLiteral("cache_version"),
+          QJsonObject {
+              { QStringLiteral("major"), SyncData::cacheVersion().first },
+              { QStringLiteral("minor"), SyncData::cacheVersion().second } } }
+    };
+    {
+        QJsonArray trackedUsersJson;
+        QJsonArray outdatedUsersJson;
+        for (const auto &user : trackedUsers) {
+            trackedUsersJson += user;
+        }
+        for (const auto &user : outdatedUsers) {
+            outdatedUsersJson += user;
+        }
+        rootObj.insert(QStringLiteral("tracked_users"), trackedUsersJson);
+        rootObj.insert(QStringLiteral("outdated_users"), outdatedUsersJson);
+        QJsonObject devicesList = toJson<QHash<QString, QHash<QString, QueryKeysJob::DeviceInformation>>>(deviceKeys);
```

Comment:
Unless you need to coerce the input parameter to a different type, you really shouldn't provide the type to `toJson()`. It's overloaded, and overloads are not template specialisations.
```suggestion
        const auto devicesList = toJson(deviceKeys);
```